File: #0091.decode-ways.cpp
class Solution {
public:
    int numDecodings(string s) {

        if (s.empty()) {
            return 0;
        }

        int size = s.size();
        vector<int> vs(size + 1);
        vs[0] = 1;
        for (int i = 1; i <= size; i ++ ) {
            if (s[i - 1] < '0' || s[i - 1] > '9') {
                return 0;
            }

            vs[i] = 0;
            if (s[i - 1] != '0') {
                vs[i] = vs[i - 1];
            }

            if (i > 1) {
                int z = (s[i-2]-'0') * 10 + s[i-1] - '0';
                if (z >= 10 && z <= 26) {
                    vs[i] += vs[i - 2];
                }
            }
        }
        return vs[size];
    }
};

File: #0091.decode-ways.js
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
    
    const length = s.length;
    const arr = Array(length + 1).fill(0);
    arr[0] = 1;
    for (let i = 1; i <= length; i++) {
        const char1 = +s.slice(i - 1, i);
        const char2 = +s.slice(i - 2, i);
        
        if (char1 > 0) {
            arr[i] = arr[i - 1];
        }
        if (char2 >= 10 && char2 <= 26) {
            arr[i] += arr[i - 2];
        }
    }    
    return arr[length];
};
